data/src/main/java/com/microsoft/azure/kusto/data/HttpClientFactory.java

```diff
@@ -0,0 +1,116 @@
+package com.microsoft.azure.kusto.data;
+
+import org.apache.http.HeaderElement;
+import org.apache.http.HeaderElementIterator;
+import org.apache.http.HttpResponse;
+import org.apache.http.conn.ConnectionKeepAliveStrategy;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.message.BasicHeaderElementIterator;
+import org.apache.http.protocol.HTTP;
+import org.apache.http.protocol.HttpContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A singleton factory of HTTP clients.
+ */
+class HttpClientFactory {
+
+    private static final Logger logger = LoggerFactory.getLogger(HttpClientFactory.class);
+
+    private HttpClientFactory() {
+    }
+
+    private static class HttpClientFactorySingleton {
+
+        private static final HttpClientFactory instance = new HttpClientFactory();
+    }
+
+    /**
+     * Returns the factory instance.
+     *
+     * @return the shared {@linkplain HttpClientFactory} instance
+     */
+    static HttpClientFactory getInstance() {
+        return HttpClientFactorySingleton.instance;
+    }
+
+    /**
+     * Creates a new Apache HTTP client.
+     *
+     * @param providedProperties custom HTTP client properties
+     * @return a new Apache HTTP client
+     */
+    CloseableHttpClient create(HttpClientProperties providedProperties) {
+        final HttpClientProperties properties = Optional.ofNullable(providedProperties)
+                .orElse(HttpClientProperties.builder().build());
+        final HttpClientBuilder httpClientBuilder = HttpClientBuilder.create()
+                .useSystemProperties()
+                .setMaxConnTotal(properties.maxConnectionTotal())
+                .setMaxConnPerRoute(properties.maxConnectionRoute())
+                .evictExpiredConnections()
+                .evictIdleConnections(properties.maxIdleTime(), TimeUnit.SECONDS);
+
+        if (properties.isKeepAlive()) {
```

Comment:
This throws if properties == null. Try
```suggestion
        if (properties.isKeepAlive() != null && properties.isKeepAlive()) {
```